hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java

```diff
@@ -932,12 +934,36 @@ public NativeQueryImplementor createNativeQuery(String sqlString, Class resultCl
 	}
 
 	private void handleNativeQueryResult(NativeQueryImplementor query, Class resultClass) {
+		boolean isObjectArray = Object[].class.equals( resultClass );
+
 		if ( Tuple.class.equals( resultClass ) ) {
 			query.setResultTransformer( new NativeQueryTupleTransformer() );
 		}
-		else {
+		else if ( resultClass.isArray() && !isObjectArray ) {
+			Class elementClass = resultClass.getComponentType();
+
+			query.setResultTransformer( new BasicTransformerAdapter() {
+				@Override
+				public Object transformTuple(Object[] tuple, String[] aliases) {
+					Object[] result = (Object[]) Array.newInstance( elementClass, tuple.length );
+					for ( int i = 0; i < tuple.length; i++ ) {
+						result[i] = elementClass.cast( tuple[i] );
+					}
+					return result;
+				}
+			} );
+		}
+		else if ( this.getFactory().getMetamodel().entityPersisters().get( resultClass.getName() ) != null ) {
 			query.addEntity( "alias1", resultClass.getName(), LockMode.READ );
 		}
+		else if ( !isObjectArray ) {
+			query.setResultTransformer( new BasicTransformerAdapter() {
+				@Override
+				public Object transformTuple(Object[] tuple, String[] aliases) {
+					return resultClass.cast( tuple[0] );
+				}
+			} );
+		}
```

Comment:
What in the `else` case? Is it legit to not do anything?

Reply:
In that case, the result is processed as `Object[]` and we don't have to do any transformation.